### PR TITLE
implement egress firewall using join switch acls in shared gateway mode

### PR DESCRIPTION
**Pull request by cathy-zhou**
Signed-off-by: Yun Zhou <yunz@nvidia.com>

**- What this PR does and why is it needed**

**- Special notes for reviewers**

**- How to verify it**

**- Description for the changelog**

### CODE REVIEW

File: go-controller/pkg/ovn/egressfirewall.go [hunk 1]
```diff
@@ -162,19 +162,10 @@ func (oc *Controller) updateEgressFirewall(oldEgressFirewall, newEgressFirewall
 	// add the new egressfirewall
 
 	updateErrors = errors.Wrapf(updateErrors, "%v", oc.addEgressFirewall(newEgressFirewall))
-	// delete policy blocking all external traffic
-	stdout, stderr, err := util.RunOVNNbctl("--data=bare", "--no-heading", "--columns=_uuid", "find",
-		"logical_router_policy", fmt.Sprintf("external-ids:egressFirewall=%s-blockAll", newEgressFirewall.Namespace))
+	// delete rules blocking all external traffic
+	err = deleteEgressFirewallRules(newEgressFirewall.Namespace + "-blockAll")
```
Comment (JacobTanenbaum) [inline go-controller/pkg/ovn/egressfirewall.go:166]: this is a great resuse of functions! could you please change the function definition?

`func deleteEgressFirewallRules(namespace string)` error -> `func deleteEgressFirewallRules(externalID string) error` to reflect that you are using the argument as the externalID and change the doc comment to that effect

File: go-controller/pkg/ovn/egressfirewall.go [hunk 2]
```diff
@@ -153,7 +153,7 @@ func (oc *Controller) updateEgressFirewall(oldEgressFirewall, newEgressFirewall
 		[]egressfirewallapi.EgressFirewallPort{},
 	)
 
-	err = createLogicalRouterPolicy(priority, match, "drop", newEgressFirewall.Namespace+"-blockAll")
+	err = createEgressFirewallRules(priority, match, "drop", newEgressFirewall.Namespace+"-blockAll")
```
Comment (JacobTanenbaum) [inline go-controller/pkg/ovn/egressfirewall.go:156]: same change request as the deleteEgressFirewallRules() function

File: go-controller/pkg/ovn/egressfirewall.go [hunk 3]
```diff
@@ -125,7 +125,7 @@ func (oc *Controller) addEgressFirewall(egressFirewall *egressfirewallapi.Egress
 		return fmt.Errorf("unable to add egress firewall policy, namespace: %s has no address set", egressFirewall.Namespace)
 	}
 
-	err = oc.addLogicalRouterPolicyToClusterRouter(nsInfo.addressSet.GetIPv4HashName(), nsInfo.addressSet.GetIPv6HashName(), egressFirewall.Namespace, egressFirewallStartPriorityInt)
+	err = oc.addEgressFirewallRules(nsInfo.addressSet.GetIPv4HashName(), nsInfo.addressSet.GetIPv6HashName(), egressFirewall.Namespace, egressFirewallStartPriorityInt)
```
Comment (JacobTanenbaum) [inline go-controller/pkg/ovn/egressfirewall.go:128]: egressFirewallStartPriorityInt comes from a list of constants for logical_router policies. Is there any issues with using the same priority on the join_switch?

**Comment by JacobTanenbaum** [comment]
Just throwing out ideas is it worth making an the rule generatioin interface and on cluster startup looking at the `config.Gateway.Mode` and choosing the correct interface? just wondering what your thoughts are this is not a required change.

File: go-controller/pkg/ovn/egressfirewall_test.go [hunk 4]
```diff
@@ -455,47 +819,47 @@ var _ = Describe("OVN test basic functions", func() {
 				ipv6source:   "",
 				destinations: []matchTarget{{matchKindV4CIDR, "1.2.3.4/32"}},
 				ports:        nil,
-				output:       `match="(ip4.dst == 1.2.3.4/32) && ip4.src == $testv4 && ip4.dst != 10.128.0.0/14"`,
+				output:       `match="(ip4.dst == 1.2.3.4/32) && ip4.src == $testv4 && inport == \"jtor-ovn_cluster_router\""`,
```
Comment (JacobTanenbaum) [inline go-controller/pkg/ovn/egressfirewall_test.go:822]: could `jtor-ovn_cluster_router` be built with the types in const.go? `\"" + types.JoinSwitchToGWRouterPrefix + types.OVNClusterRouter + "\"" `

File: go-controller/pkg/ovn/egressfirewall.go (hunk 1 above)
Comment (cathy-zhou) [reply go-controller/pkg/ovn/egressfirewall.go:166]: fixed.

File: go-controller/pkg/ovn/egressfirewall.go (hunk 2 above)
Comment (cathy-zhou) [reply go-controller/pkg/ovn/egressfirewall.go:156]: fixed

File: go-controller/pkg/ovn/egressfirewall.go (hunk 3 above)
Comment (cathy-zhou) [reply go-controller/pkg/ovn/egressfirewall.go:128]: as long as the priority does not conflict with the existing acls, I don't think there is any problem.

File: go-controller/pkg/ovn/egressfirewall_test.go (hunk 4 above)
Comment (cathy-zhou) [reply go-controller/pkg/ovn/egressfirewall_test.go]: fixed. Thanks for your comments.

**Comment by girishmg** [comment]
@trozet @dcbw @JacobTanenbaum do you guys have any more comments  on this? Thanks.

**Pull request merged**